### PR TITLE
[PROD-6800] Add option in config to set visitor cookie

### PR DESCRIPTION
**Pull request by klesgidis**
## Summary
* Adds an option (default false) to `add-visitor-id` middleware to set visitor cookie
* User can override cookie domain by setting the `config.visitor.getCookieDomain` func

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -264,6 +264,13 @@ You can find the default values below:
   "workers": {
     "retryDelay": 3600000,
     "initializationCheckDelay": 1000
+  },
+  "visitor": {
```
Comment (nikostoulas) [inline docs/configuration.md:268]: 👍

File: src/initializers/koa/add-visitor-id.ts [hunk 2]
```diff
@@ -1,25 +1,62 @@
...
+
+    if (config.visitor.setCookie && !ctx.state.visitor) {
+      ctx.state.visitor = uuid.v4();
+      if (config.visitor.secure && !ctx.cookies.secure) {
+        // Force secure cookies in context, when our ENV dictates to always use secure=true;.
+        ctx.cookies.secure = true;
+      }
+
+      let cookieDomain = domain((ctx.origin && new URL(ctx.origin).hostname) || ctx.hostname);
+      if (config.visitor.getCookieDomain && typeof config.visitor.getCookieDomain === 'function') {
```
Comment (nikostoulas) [inline src/initializers/koa/add-visitor-id.ts:42]: ❓ Undocumented field (config.visitor.getCookieDomain)?

File: docs/middleware.md [hunk 3]
```diff
@@ -23,11 +23,12 @@ Orka adds by default the following middleware:
 - errorHandler (Handles errors thrown in any middleware added afterwards)
 - [koa-compress](https://www.npmjs.com/package/koa-compress)
 - [koa2-cors](https://www.npmjs.com/package/koa2-cors)
-- addVisitorId (adds a visitorId in ctx.state.visitorId from cookie ctx.visitor.cookie)
+- addVisitorId (adds a visitorId in ctx.state.visitor from cookie ctx.visitor.cookie)
```
Comment (nikostoulas) [inline docs/middleware.md:26]: ❓ Is this a fix to the docs or did the implementation change? The first right?

File: src/initializers/koa/add-visitor-id.ts [hunk 4]
```diff
@@ -1,25 +1,62 @@
...
+        cookieDomain = config.visitor.getCookieDomain(ctx);
+      }
+
+      const options: SetOption = {
+        domain: cookieDomain,
+        maxAge: config.visitor.maxAge && ms(config.visitor.maxAge),
```
Comment (nikostoulas) [inline src/initializers/koa/add-visitor-id.ts:48]: 🙃 ms is nice! However maybe we should also document that this is used for someone to know what to add in maxAge

File: package.json [hunk 5]
```diff
@@ -97,6 +98,7 @@
     "pg": "*",
     "prettier": "^2.2.1",
     "prom-client": "*",
+    "proxyquire": "^2.1.3",
```
Comment (nikostoulas) [inline package.json:101]: 🙃 Note that we also use mock-require in this repo for the same thing.

**Review by nikostoulas — APPROVED** [review]

File: docs/middleware.md (hunk 3 above)
Comment (klesgidis) [reply docs/middleware.md:26]: Yes, this is a fix to the docs

File: package.json (hunk 5 above)
Comment (klesgidis) [reply package.json:101]: I know, but have you seen this (https://github.com/boblauer/mock-require)
Screenshot 2021-11-18 at 12 23 42 PM
?

Comment (nikostoulas) [reply package.json:101]: 😮 we should remove mock-require then at some point

**Pull request merged**